### PR TITLE
isc-dhcp: refuse to add empty DHCP ranges

### PR DESCRIPTION
**Pull request by yogo1212**
Maintainer: @pprindeville
Compile tested: no
Run tested: no

Description:

With [#12925](https://github.com/openwrt/openwrt/pull/12925), 'BROADCAST' will no longer be set if there is no local broadcast address (instead of setting the global broadcast address). Prepare for the merger but stay compatible with the old version of ipcalc.

### CODE REVIEW

**Comment by Ansuel** [comment]
@yogo1212 please bump pkg release and mention in the other pr that this should be merged first.

**Comment by yogo1212** [comment]
@Ansuel Oops!
Thanks - will do!

**Comment by yogo1212** [comment]
A second commit uses the return value of ipcalc.sh to abort generating the config from a UCI section if the range wouldn't be valid.

File: net/isc-dhcp/files/dhcpd.init [hunk 1]
```diff
@@ -350,7 +350,9 @@ gen_dhcp_subnet() {
 	echo "subnet $NETWORK netmask $NETMASK {"
 	echo " range $START $END;"
 	echo " option subnet-mask $netmask;"
-	if [ "$BROADCAST" != "0.0.0.0" ] ; then
+	# check for 0.0.0.0 until all active releases of ipcalc.sh omit it
+	# for small networks:
+	if [ -n "$BROADCAST" -a "$BROADCAST" != "0.0.0.0" ] ; then
```
Comment (pprindeville) [inline net/isc-dhcp/files/dhcpd.init:360]: Should be `[ ... ] && [ ... ]` not `[ ... -a ... ]`.

File: net/isc-dhcp/files/dhcpd.init [hunk 2]
```diff
@@ -344,6 +344,11 @@ static_hosts() {
 	config_foreach static_host_add host "$@"
 }
 
+eval_ipcalc() {
```
Comment (pprindeville) [inline net/isc-dhcp/files/dhcpd.init:347]: Why is this necessary?  Also, see PR #21410

**Review by pprindeville — CHANGES_REQUESTED** [review]

File: net/isc-dhcp/files/dhcpd.init (hunk 2 above)
Comment (yogo1212) [reply net/isc-dhcp/files/dhcpd.init:347]: This change is coming from https://github.com/openwrt/openwrt/pull/12925 and is only slightly related.
Instead of enabling modes of configuration, it is concerned with error handling.

The synopsis is: ipcalc.sh now fails if the range doesn't fit into the subnet.
The purpose of the function is to pass through the return code when using `eval`. While we're at it, don't call eval if ipcalc.sh fails.
A similar patch is pending for dnsmasq.

Comment (pprindeville) [reply net/isc-dhcp/files/dhcpd.init:368]: Hmm... why not wrap it with `export -n "$(...)"` instead?

Comment (yogo1212) [reply net/isc-dhcp/files/dhcpd.init:368]: Hmm, the quotes won't work. `IP` would be set to the entire output of `ipcalc.sh`.
xargs has no `-d` in busybox, also xargs won't work with the export built-in.

A more correct solution would be `while read LINE ; do export -n "$LINE"; done < <(ipcalc.sh 11.53.100.0 30 1 2)`.
If you want, you can grap that and pour it into a commit.
`ipcalc.sh 11.53.100.0 30 1 2 | while read LINE ; do export -n "$LINE"; done` doesn't work because the `export` runs in a subshell (because of the pipe).

Tbh, I don't think the added safety is worth the added complexity. The output of ipcalc.sh is intended for something like `eval`.

**Comment by john-tho** [comment]
Ping on this PR. @pprindeville are there any remaining problems here?
It would be great to get at least the first commit (`BROADCAST` not null) applied, so we can proceed with getting the ipcalc fixes applied in the main repo.

**Comment by yogo1212** [comment]
the `local var=$(false)` bug was fixed and the PR rebased.

**Comment by yogo1212** [comment]
@pprindeville Is there something i can help with?
We'd like to get the PR moving in the main repo because it fixes ipcalc for IPv4 prefixes >30..

**Review by pprindeville — COMMENTED** [review]
I still there might be a better solution that what `eval_ipcalc()` does, but we can circle back and fix that later.  I won't hold this up.

**Review by pprindeville — APPROVED** [review]
LGTM

**Comment by pprindeville** [comment]
See PR #22414.
